File: Tarea12_RadixSort/RadixSort.py
def counting_sort(arr, exp):
    s = len(arr)
    output_array = [0] * s
    count_array = [0] * 10 

    for j in range(s):
        idx = (arr[j] // exp) % 10 
        count_array[idx] += 1

    for j in range(1, 10):
        count_array[j] += count_array[j - 1]

    for j in range(s - 1, -1, -1):
        idx = (arr[j] // exp) % 10
        output_array[count_array[idx] - 1] = arr[j]
        count_array[idx] -= 1

    for j in range(s):
        arr[j] = output_array[j]

def radix_sort(arr):
    max1 = max(arr)
    
    exp = 1
    while max1 // exp > 0:
        counting_sort(arr, exp)
        exp *= 10

if __name__ == "__main__":
    
    arr = [171, 46, 76, 91, 803, 25, 3, 67]

    print("Arreglo antes de Radix Sort:")
    print(*(arr))

    radix_sort(arr)

    print("Después de Radix Sort:")
    print(*(arr))

File: Tarea12_RadixSort/RadixSort.java
import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;

public class RadixSort {

    private static void countingSort(List<Integer> arr, int exp) {
        int s = arr.size();
        int[] outputArray = new int[s];
        int[] countArray = new int[10]; 

        for (int j = 0; j < s; j++) {
            int idx = (arr.get(j) / exp) % 10;
            countArray[idx]++;
        }

        for (int j = 1; j < 10; j++)
            countArray[j] += countArray[j - 1];

        for (int j = s - 1; j >= 0; j--) {
            int idx = (arr.get(j) / exp) % 10;
            outputArray[countArray[idx] - 1] = arr.get(j);
            countArray[idx]--;
        }

        for (int j = 0; j < s; j++)
            arr.set(j, outputArray[j]);
    }

    private static int getMax(List<Integer> arr) {
        int max = arr.get(0);
        for (int i = 1; i < arr.size(); i++) {
            if (arr.get(i) > max) {
                max = arr.get(i);
            }
        }
        return max;
    }

    public static void radixSort(List<Integer> arr) {
        int max1 = getMax(arr);
        
        for (int exp = 1; max1 / exp > 0; exp *= 10)
            countingSort(arr, exp);
    }

    public static void main(String[] args) {
        List<Integer> arr = new ArrayList<>(Arrays.asList(171, 46, 76, 91, 803, 25, 3, 67));

        System.out.println("Arreglo antes de Radix Sort:");
        for (int n : arr) System.out.print(n + " ");
        System.out.println();

        radixSort(arr);

        System.out.println("Después de Radix Sort:");
        for (int n : arr) System.out.print(n + " ");
        System.out.println();
    }
}